change `pub enum` to `ENUM!` or `FLAGS!`

Eh2406 opened the pull request:
I often removed the numbers ware redundant

Eh2406 commented:
notes for further work on `pub enum`:

many enum's have `#[allow(unused_qualifications)]` but I did not find docs on that allow. Do you want me to convert throws as well?

sapi SpeechAudioFormatType and vss VSS_APPLICATION_LEVEL have an enum val of -1. How should I convert that to a u32?

shobjidl has DEFAULT as a val in two different enum's this is not compatible with `ENUM!`

retep998 commented:
The dummy variants for univariant enums can be removed when converting them to `ENUM!` and `FLAGS!`.
I'd prefer if the explicit discriminants weren't removed.
`#[allow(unused_qualifications)]` isn't needed when switching to `ENUM!`/`FLAGS!`.
For negative discriminants do `-1i32 as u32`.

retep998 commented:
For those enums in shobjidl, it turns out that many enums in winapi had their variant names shortened to remove the common prefix. Like for example.
```C
enum FDE_OVERWRITE_RESPONSE
    {
        FDEOR_DEFAULT	= 0,
        FDEOR_ACCEPT	= 1,
        FDEOR_REFUSE	= 2
    } 	FDE_OVERWRITE_RESPONSE;
```
It is currently represented in winapi as
```Rust
#[repr(i32)] #[derive(Clone, Copy, Debug)]
pub enum FDE_OVERWRITE_RESPONSE {
    DEFAULT = 0,
    ACCEPT = 1,
    REFUSE = 2,
}
```
What it should become:
```Rust
ENUM!{enum FDE_OVERWRITE_RESPONSE {
    FDEOR_DEFAULT = 0,
    FDEOR_ACCEPT = 1,
    FDEOR_REFUSE = 2,
}}

Eh2406 commented:
I will continue working on this when I get a chance. Thanks for all the work you are putting into this project!

Eh2406 commented:
reverted the numbers. Next commit will do the `#[allow(unused_qualifications)]`

Eh2406 commented:
ok so I converted all of the `pub enum` to `ENUM!` or `FLAGS!` but then started getting cryptic error messages. I reverted and added one file at a time to isolate the issues. The ones that can be added are in my repo, if they meet with your approval, I will squash.

Then I went back into files to figure out exactly which change is at fault.
For example D3DSHADER_INSTRUCTION_OPCODE_TYPE hits the macro recursion limit.
Tracking these down is slow. the error messages just go to the macros file. Perhaps they are for a spret pr?

retep998 commented:
For larger enums, the solution is to specify all the discriminants explicitly, otherwise it has to use macro recursion in order to figure them out. Oh how I wish Rust had stable procedural macros.

Eh2406 commented:
That is all of them! Please review, so I can squash.

retep998 commented on src/corsym.rs:
```diff
@@ -65,14 +60,12 @@ interface ISymUnmanagedBinder(ISymUnmanagedBinderVtbl): IUnknown(IUnknownVtbl) {
 );
-#[repr(i32)] #[derive(Clone, Copy, Debug)] #[allow(unused_qualifications)]
-pub enum CorSymSearchPolicyAttributes {
+ENUM!{enum CorSymSearchPolicyAttributes {
```
This should probably be `FLAGS!`

retep998 commented on src/d2d1.rs:
```diff
@@ -173,25 +145,19 @@ STRUCT!{struct D2D1_TRIANGLE {
 }}
-#[repr(i32)] #[derive(Copy, Clone, Debug)] #[allow(unused_qualifications)]
-pub enum D2D1_PATH_SEGMENT {
+ENUM!{enum D2D1_PATH_SEGMENT {
```
This should be `FLAGS!`

retep998 commented on src/d2d1.rs:
```diff
@@ -237,40 +201,30 @@ STRUCT!{struct D2D1_LAYER_PARAMETERS {
-#[repr(i32)] #[derive(Copy, Clone, Debug)] #[allow(unused_qualifications)]
-pub enum D2D1_PRESENT_OPTIONS {
+}}
+ENUM!{enum D2D1_PRESENT_OPTIONS {
```
Also `FLAGS!`.
A whole bunch of these enums here that have hexadecimal discriminants should be `FLAGS!`.

retep998 commented on src/http.rs:
```diff
@@ -516,7 +501,8 @@ pub struct HTTP_REQUEST_HEADERS {
     pub pTrailers: PHTTP_UNKNOWN_HEADER,
-    pub KnownHeaders: [HTTP_KNOWN_HEADER; HTTP_HEADER_ID::HttpHeaderRequestMaximum as usize],
+    pub KnownHeaders: [HTTP_KNOWN_HEADER; 41usize],
+    //                                    ^- FIXME: need to be HttpHeaderRequestMaximum
```
Would doing `HttpHeaderRequestMaximum.0` work here?

Eh2406 replied:
no, I get compiler error.

retep998 commented on src/processsnapshot.rs:
```diff
@@ -1,8 +1,7 @@
 //! Defines the process snapshot API
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum PSS_CAPTURE_FLAGS {
+ENUM!{enum PSS_CAPTURE_FLAGS {
```
`FLAGS!`

retep998 commented on src/processsnapshot.rs:
```diff
@@ -34,19 +32,17 @@ pub enum PSS_QUERY_INFORMATION_CLASS {
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum PSS_DUPLICATE_FLAGS {
+}}
+ENUM!{enum PSS_DUPLICATE_FLAGS {
```
`FLAGS!`

retep998 commented on src/shobjidl.rs:
```diff
@@ -439,26 +439,24 @@ pub struct IContactManagerInterop;
-}
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum SICHINTF {
+    PARENTRELATIVEPARSING = 0x80018001u32,
```
No need to have the u32 suffix here anymore.

retep998 commented on src/shobjidl.rs:
```diff
@@ -439,26 +439,24 @@ pub struct IContactManagerInterop;
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum SIGDN {
+ENUM!{enum SIGDN {
     NORMALDISPLAY = 0,
```
Variant names are missing their prefix.

retep998 commented on src/shobjidl.rs:
```diff
@@ -439,26 +439,24 @@ pub struct IContactManagerInterop;
+    PARENTRELATIVEFORUI = 0x80094001u32,
+}}
+ENUM!{enum SICHINTF {
     DISPLAY = 0,
```
More missing prefixes.

retep998 commented on src/shobjidl.rs:
```diff
@@ -515,8 +510,7 @@ interface IFileDialogEvents(IFileDialogEventsVtbl): IUnknown(IUnknownVtbl) {
 );
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum FILEOPENDIALOGOPTIONS {
+ENUM!{enum FILEOPENDIALOGOPTIONS {
```
Missing prefixes and `FLAGS!`

retep998 commented on src/shobjidl.rs:
```diff
@@ -589,13 +583,12 @@ interface IFileOpenDialog(IFileOpenDialogVtbl): IFileDialog(IFileDialogVtbl) {
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum CDCONTROLSTATEF {
+ENUM!{enum CDCONTROLSTATEF {
     INACTIVE = 0x00000000,
```
Missing prefixes.

retep998 commented on src/windowscodecs.rs:
```diff
@@ -11,35 +11,26 @@ STRUCT!{struct WICRect {
-#[repr(i32)] #[derive(Copy, Clone, Debug)] #[allow(unused_qualifications)]
-pub enum WICComponentType {
+}}
+ENUM!{enum WICComponentType {
```
`FLAGS!`

retep998 commented on src/windowscodecs.rs:
```diff
@@ -47,18 +38,14 @@ pub enum WICComponentType {
-#[allow(unused_qualifications, overflowing_literals, trivial_numeric_casts)]
-pub enum WICComponentEnumerateOptions {
+}}
+ENUM!{enum WICComponentEnumerateOptions {
```
`FLAGS!`

retep998 commented on src/windowscodecs.rs:
```diff
@@ -105,68 +87,50 @@ pub enum WICBitmapDitherType {
-#[repr(i32)] #[derive(Copy, Clone, Debug)] #[allow(unused_qualifications)]
-pub enum WICBitmapTransformOptions {
+}}
+ENUM!{enum WICBitmapTransformOptions {
```
`FLAGS!`

Eh2406 commented:
squashed and ready.

when I next have time I perhaps I can work on
```
#[repr(C)] #[derive(Clone, Copy, Debug)]
pub struct IExecuteCommandApplicationHostEnvironment;
```
I.E. the many empty struct and turning them into `STRUCT!`

retep998 commented:
Actually all those `IFoo` empty structs are supposed to be `RIDL!` interfaces, it's just that nobody took the time to flesh them out yet.

retep998 commented on src/commctrl.rs:
```diff
@@ -3433,8 +3431,7 @@ pub const CB_GETCUEBANNER: ::UINT = CBM_FIRST + 4;
 ) -> ::HRESULT>;
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum TASKDIALOG_FLAGS {
+ENUM!{enum TASKDIALOG_FLAGS {
```
`FLAGS!`

retep998 commented on src/shobjidl.rs:
```diff
@@ -515,8 +510,7 @@ interface IFileDialogEvents(IFileDialogEventsVtbl): IUnknown(IUnknownVtbl) {
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum FILEOPENDIALOGOPTIONS {
+FLAGS!{enum FILEOPENDIALOGOPTIONS {
     OVERWRITEPROMPT = 0x2,
```
Missing prefixes on variants
